### ISSUE TITLE
Page Components rendering twice on Server

### ISSUE THREAD

**Issue by NagabhushanS** [vercel/next.js#36229]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:34:37 PST 2022; root:xnu-8020.101.4~2/RELEASE_X86_64
    Binaries:
      Node: 16.13.1
      npm: 8.1.2
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.1.5
      react: 18.0.0
      react-dom: 18.0.0

warn  - Latest canary version not detected, detected: "12.1.5", newest: "12.1.6-canary.3".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue

### What browser are you using? (if relevant)

Chrome

### Describe the Bug

Every Page component renders twice on server. For, example the following code prints `TempPage Render` twice on server   in production mode (executed using npm run build; npm start)

In fact the root component defined in `_app.js` also renders twice on server.

BTW: the components render only once on client (browser) which is expected.

```
import { GetServerSideProps, NextPage } from 'next';
import React, { useEffect } from 'react';

const TempPage: NextPage<{username:string}> = ({username}) => {
  console.log("TempPage Render");

  return (
    <div>Welcome to temp {username}!</div>
  );
};

export default TempPage;

export const getServerSideProps: GetServerSideProps = async () => {
  console.log('getServerSideProps');
  
  return {
    props: {
      username: 'username',
    },
  }
};
```
Following is the output on the command line where npm run build; npm start is run:

```
getServerSideProps
TempPage Render
TempPage Render
```

### Expected Behavior

Every Page should render only once on the Server.

### To Reproduce

Create a new next app using `npx create-next-app <app-name> --ts` and then create a sample temp page using the above code.

**Comment by NagabhushanS** [comment]
I tried the above code with class component and in fact the component is being built twice.

import { GetServerSideProps, NextPage } from 'next';
import React, { useEffect } from 'react';

const TempPage: NextPage<{username:string}> = ({username}) => {
  console.log("TempPage Render");

  return (
    <div>Welcome to temp {username}!</div>
  );
};

export default TempPage;

export const getServerSideProps: GetServerSideProps = async () => {
  console.log('getServerSideProps');
  
  return {
    props: {
      username: 'username',
    },
  }
};

On npm run build and npm start, the following is output:

```
getServerSideProps
TempPage Constructor
TempPage Render
TempPage Constructor
TempPage Render
```

Seems like the component itself is being constructed twice. The same happens with _app.js component as well.

**Comment by NagabhushanS** [comment]
@ijjk Could you please provide input on this?

**Comment by HaNdTriX** [comment]
## Debugging Context:

First render is caused by (`ReactDOMServer.renderToReadableStream`): 
https://github.com/vercel/next.js/blob/canary/packages/next/server/render.tsx#L1425-L1428

Second render is caused by (`docCtx.renderPage`->`ReactDOMServer.renderToString`): 
https://github.com/vercel/next.js/blob/canary/packages/next/server/render.tsx#L1336-L1346

**Comment by NagabhushanS** [comment]
Why are both used? Why is just one not sufficient?
